src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandBinder.cs

```diff
@@ -484,123 +616,268 @@ private Expression CreateTotalCountExpression(Expression source, ExpandedReferen
+            IList<NamedPropertyExpression> includedProperties)
+        {
+            Contract.Assert(source != null);
+            Contract.Assert(structuredType != null);
+            Contract.Assert(includedProperties != null);
+
+            PropertyInfo dynamicPropertyDictionary = EdmLibHelpers.GetDynamicPropertyDictionary(structuredType, _model);
+            if (dynamicPropertyDictionary != null)
+            {
+                Expression propertyName = Expression.Constant(dynamicPropertyDictionary.Name);
+                Expression propertyValue = Expression.Property(source, dynamicPropertyDictionary.Name);
+                Expression nullablePropertyValue = ExpressionHelpers.ToNullable(propertyValue);
+                if (_settings.HandleNullPropagation == HandleNullPropagationOption.True)
+                {
+                    // source == null ? null : propertyValue
+                    propertyValue = Expression.Condition(
+                        test: Expression.Equal(source, Expression.Constant(value: null)),
+                        ifTrue: Expression.Constant(value: null, type: TypeHelper.ToNullable(propertyValue.Type)),
+                        ifFalse: nullablePropertyValue);
+                }
+                else
+                {
+                    propertyValue = nullablePropertyValue;
+                }
+
+                includedProperties.Add(new NamedPropertyExpression(propertyName, propertyValue));
+            }
         }
 
-        private static SelectExpandClause GetOrCreateSelectExpandClause(KeyValuePair<IEdmNavigationProperty, ExpandedReferenceSelectItem> propertyToExpand)
+        private static SelectExpandClause GetOrCreateSelectExpandClause(IEdmNavigationProperty navigationProperty, ExpandedReferenceSelectItem expandedItem)
         {
             // for normal $expand=....
-            ExpandedNavigationSelectItem expandNavigationSelectItem = propertyToExpand.Value as ExpandedNavigationSelectItem;
+            ExpandedNavigationSelectItem expandNavigationSelectItem = expandedItem as ExpandedNavigationSelectItem;
             if (expandNavigationSelectItem != null)
             {
                 return expandNavigationSelectItem.SelectAndExpand;
             }
 
-            // for $expand=..../$ref, just includes the keys properties.
+            // for $expand=..../$ref, just includes the keys properties
             IList<SelectItem> selectItems = new List<SelectItem>();
-            foreach (IEdmStructuralProperty keyProperty in propertyToExpand.Key.ToEntityType().Key())
+            foreach (IEdmStructuralProperty keyProperty in navigationProperty.ToEntityType().Key())
             {
                 selectItems.Add(new PathSelectItem(new ODataSelectPath(new PropertySegment(keyProperty))));
             }
 
+            // TODO: Sam, it seems we can only create a "SelectAll==false" and leave the "SelectItems as empty",
+            // because the keys will be included auto-selected.
```

Comment:
Is this something to be fixed before release? Or should you open an issue to track it?

Reply:
No. It seems it's a question for me in the implementation. Now, we can remove this.

---
In reply to: [355633970](https://github.com/OData/WebApi/pull/1941#discussion_r355633970) [](ancestors = 355633970)